huaweicloud/services/modelarts/resource_huaweicloud_modelarts_authorization.go

```diff
@@ -0,0 +1,288 @@
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Agency name.`,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^(modelarts_agency)$|^(ma_agency_.+)$`),
+					"The agency name can be modelarts_agency or prefixed with ma_agency_"),
+			},
+			"user_name": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `User Name.`,
+			},
+		},
+	}
+}
+
+func resourceModelArtsAuthorizationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	// createAgency: create a ModelArts agency.
+	var (
+		createAgencyHttpUrl = "v2/{project_id}/agency"
+		createAgencyProduct = "modelarts"
+	)
+	createAgencyClient, err := cfg.NewServiceClient(createAgencyProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating ModelArts Client: %s", err)
+	}
+
+	createAgencyPath := createAgencyClient.Endpoint + createAgencyHttpUrl
+	createAgencyPath = strings.ReplaceAll(createAgencyPath, "{project_id}", createAgencyClient.ProjectID)
+
+	createAgencyOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+		MoreHeaders: map[string]string{"Content-Type": "application/json"},
+	}
+
+	createAgencyOpt.JSONBody = utils.RemoveNil(buildCreateAgencyBodyParams(d))
+	_, err = createAgencyClient.Request("POST", createAgencyPath, &createAgencyOpt)
+	if err != nil && parseAgencyErrorIgnoreConflict(err) != nil {
+		return diag.Errorf("error creating ModelArts agency: %s", err)
+	}
+
+	// createAuth: create a ModelArts authorization.
+	var (
+		createAuthHttpUrl = "v2/{project_id}/authorizations"
+		createAuthProduct = "modelarts"
+	)
+	createAuthClient, err := cfg.NewServiceClient(createAuthProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating ModelArts Client: %s", err)
+	}
+
+	createAuthPath := createAuthClient.Endpoint + createAuthHttpUrl
+	createAuthPath = strings.ReplaceAll(createAuthPath, "{project_id}", createAuthClient.ProjectID)
+
+	createAuthOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+		MoreHeaders: map[string]string{"Content-Type": "application/json"},
+	}
+
+	createAuthOpt.JSONBody = utils.RemoveNil(buildCreateAuthBodyParams(d))
+	_, err = createAuthClient.Request("POST", createAuthPath, &createAuthOpt)
+	if err != nil {
+		return diag.Errorf("error creating ModelArts authorization: %s", err)
+	}
+
+	d.SetId(d.Get("user_id").(string))
+
+	return resourceModelArtsAuthorizationRead(ctx, d, meta)
+}
+
+func buildCreateAgencyBodyParams(d *schema.ResourceData) map[string]interface{} {
+	agencyNameSuffix := "" // when agency_name is modelarts_agency, set empty
+
+	agencyName := d.Get("agency_name").(string)
+	// when agency_name is prefixed with 'ma_agency_', take the suffix
+	if strings.HasPrefix(agencyName, "ma_agency_") {
+		agencyNameSuffix = strings.TrimPrefix(agencyName, "ma_agency_")
+	}
```

Comment:
I think we need to preCheck `agency_name` value. If it is not equal `modelarts_agency` or not start with `ma_agency_`, return an invalid error.